File: packages/stim-proto/src/reply.ts
import type { ConversationId, MessageId, ReplyId } from "./ids.js";

export type ReplyStatus = "pending" | "streaming" | "completed" | "failed";

export interface ReplyHandle {
  reply_id: ReplyId;
  conversation_id: ConversationId;
  message_id: MessageId;
  status: ReplyStatus;
}

export interface ReplyFailure {
  code: string;
  message: string;
}

export interface ReplySnapshot {
  reply_id: ReplyId;
  conversation_id: ConversationId;
  message_id: MessageId;
  status: ReplyStatus;
  output_text: string;
  error?: ReplyFailure;
}

export type ReplyEventKind =
  | {
      type: "output_text_delta";
      delta: string;
    }
  | {
      type: "completed";
    }
  | {
      type: "failed";
      error: ReplyFailure;
    };

export interface ReplyEvent {
  reply_id: ReplyId;
  sequence: number;
  event: ReplyEventKind;
}

File: crates/stim-proto/src/reply.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::{ConversationId, MessageId, ReplyId};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ReplyStatus {
    Pending,
    Streaming,
    Completed,
    Failed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct ReplyHandle {
    pub reply_id: ReplyId,
    pub conversation_id: ConversationId,
    pub message_id: MessageId,
    pub status: ReplyStatus,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct ReplyFailure {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct ReplySnapshot {
    pub reply_id: ReplyId,
    pub conversation_id: ConversationId,
    pub message_id: MessageId,
    pub status: ReplyStatus,
    pub output_text: String,
    pub error: Option<ReplyFailure>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ReplyEventKind {
    OutputTextDelta { delta: String },
    Completed,
    Failed { error: ReplyFailure },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct ReplyEvent {
    pub reply_id: ReplyId,
    pub sequence: u64,
    pub event: ReplyEventKind,
}
